code/modules/tgui/tgui.dm

```diff
@@ -0,0 +1,363 @@
+/**
+ * tgui
+ *
+ * /tg/station user interface library
+ */
+
+/**
+ * tgui datum (represents a UI).
+ */
+/datum/tgui
+	/// The mob who opened/is using the UI.
+	var/mob/user
+	/// The object which owns the UI.
+	var/datum/src_object
+	/// The title of te UI.
+	var/title
+	/// The ui_key of the UI. This allows multiple UIs for one src_object.
+	var/ui_key
+	/// The window_id for browse() and onclose().
+	var/window_id
+	/// The window width.
+	var/width = 0
+	/// The window height
+	var/height = 0
+	/// The interface (template) to be used for this UI.
+	var/interface
+	/// Update the UI every MC tick.
+	var/autoupdate = TRUE
+	/// If the UI has been initialized yet.
+	var/initialized = FALSE
+	/// The data (and datastructure) used to initialize the UI.
+	var/list/initial_data
+	/// The static data used to initialize the UI.
+	var/list/initial_static_data
+	/// Holder for the json string, that is sent during the initial update
+	var/_initial_update
+	/// The status/visibility of the UI.
+	var/status = STATUS_INTERACTIVE
+	/// Topic state used to determine status/interactability.
+	var/datum/tgui_state/state = null
+	/// The parent UI.
+	var/datum/tgui/master_ui
+	/// Children of this UI.
+	var/list/datum/tgui/children = list()
+
+/**
+ * public
+ *
+ * Create a new UI.
+ *
+ * required user mob The mob who opened/is using the UI.
+ * required src_object datum The object or datum which owns the UI.
+ * required ui_key string The ui_key of the UI.
+ * required interface string The interface used to render the UI.
+ * optional title string The title of the UI.
+ * optional width int The window width.
+ * optional height int The window height.
+ * optional master_ui datum/tgui The parent UI.
+ * optional state datum/ui_state The state used to determine status.
+ *
+ * return datum/tgui The requested UI.
+ */
+/datum/tgui/New(mob/user, datum/src_object, ui_key, interface, title, width = 0, height = 0, datum/tgui/master_ui = null, datum/tgui_state/state = GLOB.tgui_default_state)
+	src.user = user
+	src.src_object = src_object
+	src.ui_key = ui_key
+	src.window_id = "[src_object.UID()]-[ui_key]"
+	src.interface = interface
+
+	if(title)
+		src.title = sanitize(title)
+	if(width)
+		src.width = width
+	if(height)
+		src.height = height
+
+	src.master_ui = master_ui
+	if(master_ui)
+		master_ui.children += src
+	src.state = state
+
+	var/datum/asset/tgui_assets = get_asset_datum(/datum/asset/simple/tgui)
+	var/datum/asset/fa = get_asset_datum(/datum/asset/simple/fontawesome)
```

Comment:
I believe TG combined this into:

`var/datum/asset/assets = get_asset_datum(/datum/asset/group/tgui)`

Reply:
We dont have asset groups so I did it this way